### PR TITLE
Added Automatic EFS Setup Script

### PR DESCRIPTION
**Pull request by AlexandreBrown**
**Description of your changes:**
This PR adds a script that completely automates EFS setup with dynamic provisioning.  
It handles cases where resources were already created as well, in which case it skips the part that was already done.

## Script In Action  
https://user-images.githubusercontent.com/26939775/153107194-1864d838-13ac-449b-9c44-bdef0bc03e75.mp4

## Example when resources were already created
https://user-images.githubusercontent.com/26939775/153107206-2d879a4f-52f1-4804-9513-e49e79609641.mp4

**Checklist:**
- [x] Dynamic provisioning
- [x] Convert to Python Script
- [x] End to end test
- [x] Unit tests pass:
  **Make sure you have installed kustomize == 3.2.1**
    1. `make generate-changed-only`
    2. `make test`

### CODE REVIEW

**Comment by surajkota** [comment]
@AlexandreBrown do mind moving this scripts to python similar to the cognito setup? 

I see the following disadvantages with using shell vs python:
* Poor support for type checking/IDE code completion
* Relies on Bash error handling - any error may cause early termination (possibly leaving resources stranded) if using `set -e`
* Can be affected by system environment variables
* Not agnostic to operating system (e.g. [issue with sed command](https://github.com/awslabs/kubeflow-manifests/pull/78#discussion_r799153838))
* Logging is difficult to follow (especially with set -x)
* Limitations of bash scripting language
    * Unable to represent objects - only primitive values
    * Global/local scope variables are unreliable and confusing
    * Unable to return anything other than exit codes from bash functions

I say python because we want to have these scripts run as part of our e2e tests so they stay up to date and since the e2e framework is written in python it will be easier to integrate

**Comment by AlexandreBrown** [comment]
> @AlexandreBrown do mind moving this scripts to python similar to the cognito setup? Since this PR is early stages of development
> 
> I see the following disadvantages with using shell vs python:
> 
> * Poor support for type checking/IDE code completion
> * Relies on Bash error handling - any error may cause early termination (possibly leaving resources stranded) if using `set -e`
> * Can be affected by system environment variables
> * Not agnostic to operating system (e.g. [issue with sed command](https://github.com/awslabs/kubeflow-manifests/pull/78#discussion_r799153838))
> * Logging is difficult to follow (especially with set -x)
> * Limitations of bash scripting language
>   
>   * Unable to represent objects - only primitive values
>   * Global/local scope variables are unreliable and confusing
>   * Unable to return anything other than exit codes from bash functions
> 
> I say python because we want to have these scripts run as part of our e2e tests so they stay up to date and since the e2e framework is written in python it will be easier to integrate

@surajkota  Sure thing, I can try to convert it to a Python script. I can also do the same for https://github.com/awslabs/kubeflow-manifests/pull/78

**Comment by AlexandreBrown** [comment]
@surajkota I've converted the bash script to a python script, also made some further improvement to the doc, more noticeably this section : https://github.com/AlexandreBrown/kubeflow-manifests/blob/feat/auto-efs-setup/distributions/aws/examples/storage/efs/README.md#30-using-efs-storage-in-kubeflow but also the beginning where we explain the dependencies for the script.  
I also improved the script to wait for the EFS file system to become `available` and same thing for the mount target to avoid errors when creating the mount target too quick (it could fail otherwise).  
  
I tested the changes on a test cluster from the v1.3 branch as well.  
  
Let me know what you think.

File: distributions/aws/examples/storage/efs/auto-efs-setup.py [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+        create_efs_file_system()
+    else:
+        print(f"Skipping EFS file system creation, file system '{EFS_FILE_SYSTEM_NAME}' already exists!")
+
+
+def does_need_to_create_efs_file_system() -> bool:
+    efs_client = get_efs_client()
+    efs_fil_systems = efs_client.describe_file_systems()["FileSystems"]
+
+    return not any(file_system["Name"] == EFS_FILE_SYSTEM_NAME for file_system in efs_fil_systems)
```
Comment (akartsky) [inline distributions/aws/examples/storage/efs/auto-efs-setup.py:226]: Typo: `efs_fil_systems` should be `efs_file_systems`

Comment (akartsky) [inline distributions/aws/examples/storage/efs/auto-efs-setup.py:226]: Bug: `file_system["Name"]` errors out `KeyError: 'Name'`
if there are no existing file systems

**Review by akartsky — APPROVED** [review]

File: distributions/aws/examples/storage/efs/README.md [hunk 2]
```diff
@@ -113,9 +133,39 @@ efs-claim   Bound    efs-pv   5Gi        RWX            efs-sc         5d16h
...
+```
+### 3.2 Creating and using EFS volumes as workspace or data volume for a notebook [Dynamic Provisioning]
+
+The following example works if you used the automatic setup with dynamic provisioning.
+
+Here is an example that illustrates how to create a PVC for your dataset in `ReadWriteMany` mode meaning it can be used by many notebooks at the same time as well as how to create a notebook with a workspace volume for the notebook data and how to specify that you want to use your dataset volume as data volume.
+Note that both of these volumes are created under the storage class `efs-sc` which represents the EFS storage class created earlier.
+  
+
+https://user-images.githubusercontent.com/26939775/153103745-70f93ac5-88f3-4387-b40d-b585fca80af4.mp4
```
Comment (mbaijal) [inline distributions/aws/examples/storage/efs/README.md:161]: While I really like this screen capture linked, we need to keep the format for the static and dynamic provisioning the same - for now could you write as text please ? ok with adding the video as well, but ideally we should then include it for both static and dynamic.

**Review by mbaijal — CHANGES_REQUESTED** [review]

File: distributions/aws/examples/storage/efs/README.md (hunk 2 above)
Comment (AlexandreBrown) [reply distributions/aws/examples/storage/efs/README.md:161]: Do you think we could make another PR to improve the doc for static provisioning Instead ? 

We could make a quick video showing how to create a PV and PVC etc.

File: distributions/aws/examples/storage/efs/README.md [hunk 3]
```diff
@@ -66,35 +88,33 @@ eksctl create iamserviceaccount \
...
-## 4.0 Create an Instance of the EFS Filesystem
+#### 3. Manually Create an Instance of the EFS Filesystem
 Please refer to the official [AWS EFS CSI Document](https://docs.aws.amazon.com/eks/latest/userguide/efs-csi.html#efs-create-filesystem) for detailed instructions on creating an EFS filesystem. 
 
 Note: For this README, we have assumed that you are creating your EFS Filesystem in the same VPC as your EKS Cluster. 
 
-## 5.0 Using EFS Storage in Kubeflow
-
-## 5.1 Static Provisioning
+#### 4. Static Provisioning
```
[the remark this answers is not in the document]
Comment (AlexandreBrown) [reply distributions/aws/examples/storage/efs/README.md:116]: Done!

**Review by mbaijal — APPROVED** [review]
Thanks for your contribution @AlexandreBrown, this is great!

**Pull request merged**